include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl

```diff
@@ -6,425 +6,762 @@
+		if (k < 0.0)
+			return T(0.0);
+		
+		return eta * I - (eta * dot_helper<T>::__call(N, I) + std::sqrt(k)) * N;
+	}
 };
-template<typename Integer>
-using find_lsb_return_type = find_msb_return_type<Integer>;
 
-template<typename T NBL_STRUCT_CONSTRAINABLE>
-struct bitReverse_helper;
+#endif // C++ only specializations
 
-template<typename Integer>
-NBL_PARTIAL_REQ_TOP(hlsl::is_integral_v<Integer> && hlsl::is_scalar_v<Integer>)
-struct bitReverse_helper<Integer NBL_PARTIAL_REQ_BOT(hlsl::is_integral_v<Integer>&& hlsl::is_scalar_v<Integer>) >
+// C++ and HLSL specializations
+
+template<typename T, uint16_t Bits>
+NBL_PARTIAL_REQ_TOP(concepts::UnsignedIntegralScalar<T> && (Bits <= sizeof(T) * 8))
+struct bitReverseAs_helper<T, Bits NBL_PARTIAL_REQ_BOT(concepts::UnsignedIntegralScalar<T> && (Bits <= sizeof(T) * 8)) >
 {
-	static inline Integer __call(NBL_CONST_REF_ARG(Integer) val)
+	static T __call(NBL_CONST_REF_ARG(T) val)
 	{
-#ifdef __HLSL_VERSION
-		return spirv::bitReverse(val);
-#else
-		return glm::bitfieldReverse(val);
-#endif
+		return bitReverse_helper<T>::__call(val) >> promote<T, scalar_type_t<T> >(scalar_type_t <T>(sizeof(T) * 8 - Bits));
+	}
+
+	static T __call(NBL_CONST_REF_ARG(T) val, uint16_t bits)
+	{
+		return bitReverse_helper<T>::__call(val) >> promote<T, scalar_type_t<T> >(scalar_type_t <T>(sizeof(T) * 8 - bits));
 	}
 };
 
-template<typename Vector>
-NBL_PARTIAL_REQ_TOP(hlsl::is_vector_v<Vector>)
-struct bitReverse_helper<Vector NBL_PARTIAL_REQ_BOT(hlsl::is_integral_v<Vector> && hlsl::is_vector_v<Vector>) >
+template<typename Vectorial>
+NBL_PARTIAL_REQ_TOP(concepts::Vectorial<Vectorial>)
+struct dot_helper<Vectorial NBL_PARTIAL_REQ_BOT(concepts::Vectorial<Vectorial>) >
+{
+	using scalar_type = typename vector_traits<Vectorial>::scalar_type;
+
+	static inline scalar_type __call(NBL_CONST_REF_ARG(Vectorial) lhs, NBL_CONST_REF_ARG(Vectorial) rhs)
+	{
+		static const uint32_t ArrayDim = vector_traits<Vectorial>::Dimension;
+		static array_get<Vectorial, scalar_type> getter;
+
+		scalar_type retval = getter(lhs, 0) * getter(rhs, 0);
+		for (uint32_t i = 1; i < ArrayDim; ++i)
+			retval = retval + getter(lhs, i) * getter(rhs, i);
+
+		return retval;
+	}
+};
+template<typename FloatingPointLikeVectorial>
+NBL_PARTIAL_REQ_TOP(concepts::FloatingPointLikeVectorial<FloatingPointLikeVectorial> && (vector_traits<FloatingPointLikeVectorial>::Dimension == 3))
+struct cross_helper<FloatingPointLikeVectorial NBL_PARTIAL_REQ_BOT(concepts::FloatingPointLikeVectorial<FloatingPointLikeVectorial> && (vector_traits<FloatingPointLikeVectorial>::Dimension == 3)) >
 {
-	static Vector __call(NBL_CONST_REF_ARG(Vector) vec)
+	static FloatingPointLikeVectorial __call(NBL_CONST_REF_ARG(FloatingPointLikeVectorial) lhs, NBL_CONST_REF_ARG(FloatingPointLikeVectorial) rhs)
 	{
 #ifdef __HLSL_VERSION
-		return spirv::bitReverse(vec);
+		return spirv::cross(lhs, rhs);
```

Comment:
you will try to call cross intrinsic on an emulated vector, you need the natural spir-v partial spec. Also pass explicit template param.

Then this one can service all vectorials in C++ and in HLSL only Vectorials that are not vectors.